pkg/operator/latencyprofilecontroller/controller.go

```diff
@@ -0,0 +1,233 @@
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	// static pod revision config maps used by installer controller to track
+	// configs across different revisions
+	revisionConfigMapName = "config"
+	revisionConfigMapKey  = "config.yaml"
+
+	// set of reasons used for updating status
+	reasonLatencyProfileUpdated         = "ProfileUpdated"
+	reasonLatencyProfileUpdateTriggered = "ProfileUpdateTriggered"
+	reasonLatencyProfileEmpty           = "ProfileEmpty"
+
+	// prefix used with staus types
+	wlpPrefix = "WorkerLatencyProfile"
+)
+
+// LatencyProfileController either instantly via the informers
+// or periodically via resync, lists the config/v1/node object
+// and fetches the worker latency profile applied on the cluster which is used to
+// updates the status of respective operator resource that uses this controller.
+// The current state of the operand is by watching the configs applied
+// to different static pod revisions that are active and uses the information to
+// update status as either progressing or completed or degraded.
+
+type LatencyProfileController struct {
+	operatorClient          v1helpers.StaticPodOperatorClient
+	targetNamespace         string
+	configMapLister         listersv1.ConfigMapNamespaceLister
+	configNodeLister        listerv1.NodeLister
+	latencyProfileArgValMap map[apiconfigv1.WorkerLatencyProfileType][]nodeobserver.LatencyProfileArgVal
+}
+
+func NewLatencyProfileController(
+	operatorClient v1helpers.StaticPodOperatorClient,
+	targetNamespace string,
+	latencyProfileArgValMap map[apiconfigv1.WorkerLatencyProfileType][]nodeobserver.LatencyProfileArgVal,
+	nodeInformer configv1informers.NodeInformer,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	eventRecorder events.Recorder,
+) factory.Controller {
+
+	ret := &LatencyProfileController{
+		operatorClient:          operatorClient,
+		targetNamespace:         targetNamespace,
+		latencyProfileArgValMap: latencyProfileArgValMap,
+		configMapLister:         kubeInformersForNamespaces.ConfigMapLister().ConfigMaps(targetNamespace),
+		configNodeLister:        nodeInformer.Lister(),
+	}
+
+	return factory.New().WithInformers(
+		// this is for our general configuration input and our status output in case another actor changes it
+		operatorClient.Informer(),
+
+		// We use nodeInformer for observing current worker latency profile
+		nodeInformer.Informer(),
+
+		// for configmaps of operator client target namespace
+		kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer(),
+	).ResyncEvery(time.Minute).WithSync(ret.sync).WithSyncDegradedOnError(operatorClient).ToController(
+		"WorkerLatencyProfile",
+		eventRecorder.WithComponentSuffix("latency-profile-controller"),
+	)
+}
+
+func (c *LatencyProfileController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+
+	// Collect the current latency profile
+	configNodeObj, err := c.configNodeLister.Get("cluster")
```

Comment:
Sounds good to me. Lemme try this approach and code it up, to test it through.

Reply:
np, hth, just remember that the current state is operator specific, we could code it as a function provided by the operators.

Reply:
Yeah.

- Controller: So using `type MatchProfileRevisionConfigsFunc func(profile configv1.WorkerLatencyProfileType, revisions []int32) (match bool, err error)` function signature as controller init inputs such that operator specific profile matching logic across different latency profile revisions can be independent.

- Config Observer: using the same `conditionFn func() (bool, error)` which can hold operator specific logic

Reply:
Shouldn't we at least try to persist the old value when gating the updates? I think we are causing following problems with such a simple gating.

let's start with following state

```
node1: low
node2: low
node3: low
```

update to medium

```
node1: low
node2: medium
node3: medium
```

update some other value

```
node1: low
node2: medium
node3: default
```

In this case we would get 3 different latency profiles and that is not desirable from what I understand. Also other combinations are possible.

Reply:
The current implementation (as per https://github.com/openshift/library-go/pull/1364#discussion_r866696610) i.e. when the config observer is invoked as follows is able to tackle the race condition which we were initially facing.
```
nodeobserver.NewLatencyProfileObserver(
	latencyConfigs,
	nodeobserver.NewSuppressConfigUpdateUntilSameProfileFunc(
		operatorClient, configMapLister, latencyConfigs,
	),
),
```

Essentially, we would want to suppress further config updates (i.e. further updates to the reqd. arg vals) even if the user changed the latency profile on the cluster until the last latency profile is attained completely by all replicas of operand for current operator as this helps to ensure cluster stability.

@atiratree the observer when invoked in the said manner should not result in `node1: low, node2: medium, node: default` atleast. Also, in transition a mix of two latency profiles eg. `node1: low, node2: medium, node3: medium` is fine and our cluster would still remain stable until replicas are not in 2+ unique profiles. All we want is that our config observer should send new values only once it's confirmed that a single latency profile have been achieved which helps in stability.

Plus, regarding your suggestion on keeping track of the internal state and persist it somewhere before allowing jump to another profile is a reasonable ask, but me and @p0lyn0mial after significant brainstorming found it even harder to keep track of internal states and avoid race, until the fact that this idea just worked out:
- compare the current latency profile among replicas
- allow the observer to transition (desired -> current) to a new value only if the comparison checked out okay
It is simpler in implementation and does what we wanna do quite well, essentially instead of comparing internal state with desired state we're now just saying that our config observer will not update the values until of all of the replicas are on the same internal state (i.e. have same arg val pairs irrespective of which profile which values).

@p0lyn0mial I tried out the entire approach on a real cluster yesterday, seems like it works out quite well, I tried to check for different profile updates. Eg. I tried switching latency profile from `Default` -> `Medium` and then again to `Default` in a very short while, the args for Default were updated back again only after all the replicas achieved the revision which contained medium args. So the config observer got suppressed for the period until all replicas where able to achieve medium profile.

Reply:
@swghosh I am sorry for ringing false alarms, I simply misunderstood how ConfigObserver controller works. Now that I have tested it properly, the feature and profile switching seems to work well. I was not able to achieve more than 2 simultaneous profiles.

Reply:
No problem @atiratree, thanks for discussing it through and glad that the functionality is well tested now to achieve desired outcome. :)